Lib/idlelib/zoomheight.py

```diff
@@ -2,42 +2,121 @@
 
 import re
 import sys
+import tkinter
 
-from idlelib import macosx
+
+class WmInfoGatheringError(Exception):
+    pass
 
 
 class ZoomHeight:
 
     def __init__(self, editwin):
         self.editwin = editwin
 
+        # Cached values for maximized window dimensions, one for each set
+        # of screen dimensions.
+        self._max_height_and_y_coords = {}
+
+    @property
+    def top(self):
+        return self.editwin.top
+
     def zoom_height_event(self, event=None):
-        top = self.editwin.top
-        zoomed = zoom_height(top)
-        menu_status = 'Restore' if zoomed else 'Zoom'
-        self.editwin.update_menu_label(menu='options', index='* Height',
-                                       label=f'{menu_status} Height')
+        zoomed = self.zoom_height()
+
+        if zoomed is not None:
+            menu_status = 'Restore' if zoomed else 'Zoom'
+            self.editwin.update_menu_label(menu='options', index='* Height',
+                                           label=f'{menu_status} Height')
+
         return "break"
 
+    def zoom_height(self):
+        width, height, x, y = get_window_geometry(self.top)
```

Comment:
I strongly prefer having multiple uses of `self.top` reduced to `top`, as it wap before, after adding `top = self.top`. I will do it if you prefer.  Ditto for the next function.